### PR TITLE
Pm/device - initial fixes

### PR DESCRIPTION
**Pull request by ceolin**
- Do not wake up device unnecessarily
- Avoid unnecessary casts
- Keep subsystem only states private

### CODE REVIEW

**Comment by erwango** [comment]
Thanks @ceolin for working on this!
Some improvements, but i'm still seeing strange behavior when running `samples/boards/stm32/power_mgmt/blinky`
The LED (on GPIO G) is blinking, but some clocks are still unexpectedly resumed:

Breakpoint 3, gpio_stm32_clock_request (dev=0x200000ac <__device_dts_ord_19>, on=on@entry=true) 
> Switch on GPIO G on gpio_pin_configure_dt entry   -> OK
Breakpoint 3, gpio_stm32_clock_request (dev=0x200000ac <__device_dts_ord_19>, on=on@entry=false)
> Switch on GPIO G off on pm_suspend               -> I'm not certain it should happen in DEVICE_IDLE (debatable)
Breakpoint 3, gpio_stm32_clock_request (dev=0x2000016c <__device_dts_ord_137>, on=on@entry=true)
> Switch on GPIO A on pm_resume                    -> should not happen
Breakpoint 3, gpio_stm32_clock_request (dev=0x2000016c <__device_dts_ord_137>, on=on@entry=true)
> Switch on GPIO A on pm_resume                    -> should really not happen
Breakpoint 3, gpio_stm32_clock_request (dev=0x200000ac <__device_dts_ord_19>, on=on@entry=true)
> Switch on GPIO G on pm_resume                    -> should really not happen
Breakpoint 3, gpio_stm32_clock_request (dev=0x200000ac <__device_dts_ord_19>, on=on@entry=false)
> Switch on GPIO G on pm_resume 
Breakpoint 3, gpio_stm32_clock_request (dev=0x200000ac <__device_dts_ord_19>, on=on@entry=true)
> Switch on GPIO G on pm_resume 
Breakpoint 3, gpio_stm32_clock_request (dev=0x2000014c <__device_dts_ord_138>, on=on@entry=true)
> Switch on GPIO B on pm_resume                    -> should not happen 
Breakpoint 3, gpio_stm32_clock_request (dev=0x2000016c <__device_dts_ord_137>, on=on@entry=true)
> Switch on GPIO A on pm_resume                    -> should not happen 
Breakpoint 3, gpio_stm32_clock_request (dev=0x2000014c <__device_dts_ord_138>, on=on@entry=true)
> Switch on GPIO B on pm_resume                    -> should not happen 
Breakpoint 3, gpio_stm32_clock_request (dev=0x2000016c <__device_dts_ord_137>, on=on@entry=true)
> Switch on GPIO A on pm_resume                    -> should not happen

Two issues:
- Still some wake up required on non used devices (GPIO A, GPIO B).
  - Not expected
  - Additionally, we could see 2 "clock on" w/o "clock off"  in between -> potential issue in the state machine
- pm_suspend and pm_resumed are invoqued on GPIO G
  - I'm not sure it should happen as device is supposed to handle its own state based on user request

**Comment by gmarull** [comment]
Note: added myself to the review list, GH seems to be doing something weird with the reviewers list.

File: drivers/gpio/gpio_stm32.c [hunk 1]
```diff
@@ -561,6 +575,63 @@ static const struct gpio_driver_api gpio_stm32_driver = {
...
+	struct gpio_stm32_data *data = dev->data;
+	int ret = 0;
+
+	if (new_state == DEVICE_PM_ACTIVE_STATE) {
+		ret = gpio_stm32_clock_request(dev, true);
+	} else if (new_state == DEVICE_PM_SUSPEND_STATE) {
+		ret = gpio_stm32_clock_request(dev, false);
+	}
+
+	data->power_state = new_state;
```
Comment (gmarull) [inline drivers/gpio/gpio_stm32.c:598]: state will be changed regardless of `ret` value

File: subsys/power/device.c [hunk 2]
```diff
@@ -60,26 +60,65 @@ static device_idx_t num_pm;
...
+	/*
+	 * If the device is currently powered off or the request was
+	 * to go to the same state, just ignore it.
+	 */
+	if ((current_state == DEVICE_PM_OFF_STATE) ||
+			(current_state == state)) {
+		return false;
+	}
+
+	return true;
```
Comment (gmarull) [inline subsys/power/device.c:84]: nit:

could be replaced by `return ((current_state != DEVICE_PM_OFF_STATE) && (current_state != state))`

File: subsys/power/device.c [hunk 3]
```diff
@@ -60,26 +60,65 @@ static device_idx_t num_pm;
...
+			/*
+			 * Don't bother the device if it is currently
+			 * in the right state.
+			 */
+			rc = device_set_power_state(dev, state, NULL, NULL);
+			if ((rc != -ENOTSUP) && (rc != 0)) {
+				LOG_DBG("%s did not enter %s state: %d",
+					dev->name, device_pm_state_str(state),
+					rc);
+				return rc;
```
Comment (gmarull) [inline subsys/power/device.c:108]: in case of error, some devices will be left on `state` while some others will be kept on the current state. I wonder if this function should *undo* the state change in case of error.

File: drivers/display/display_st7789v.c [hunk 4]
```diff
@@ -409,14 +409,14 @@ static void st7789v_enter_sleep(struct st7789v_data *data)
 }
 
 static int st7789v_pm_control(const struct device *dev, uint32_t ctrl_command,
-				 void *context, device_pm_cb cb, void *arg)
+				 uint32_t *state, device_pm_cb cb, void *arg)
```
Comment (gmarull) [inline drivers/display/display_st7789v.c:412]: that's a nice improvement. I wonder if states could be represented by an `enum`.

File: subsys/power/device_pm.c [hunk 5]
```diff
@@ -177,6 +190,9 @@ void device_pm_enable(const struct device *dev)
 
 void device_pm_disable(const struct device *dev)
 {
+	__ASSERT(k_is_pre_kernel() == false, "Device should not be disabled"
+		 "before kernel be initialized");
```
Comment (gmarull) [inline subsys/power/device_pm.c:194]: nit:
`before kernel is initialized`.

File: subsys/power/device_pm.c [hunk 6]
```diff
@@ -112,6 +112,10 @@ static int device_pm_request(const struct device *dev,
 		}
 	}
 
+	if (k_is_pre_kernel()) {
```
Comment (gmarull) [inline subsys/power/device_pm.c:115]: I assume these changes imply that `pm_work_handler` will be never run in pre_kernel stage. Let's discuss that later more in detail.

File: drivers/pinmux/stm32/pinmux_stm32.c [hunk 7]
```diff
@@ -151,8 +151,8 @@ int stm32_dt_pinctrl_configure(const struct soc_gpio_pinctrl *pinctrl,
 #ifdef CONFIG_PM_DEVICE_IDLE
 		/* FIXME: Use of device_pm clock mgmt function */
 		/* breaks correct gpio driver behavior */
```
Comment (gmarull) [inline drivers/pinmux/stm32/pinmux_stm32.c:153]: `FIXME` comments likely need to go away now

File: drivers/display/display_st7789v.c (hunk 4 above)
Comment (ceolin) [reply drivers/display/display_st7789v.c:412]: yep, I will work on states like we did to system power management. Just changed this because this signature was driving me crazy, every time I looked it I thought that context was some private data passing across callbacks but was just the state :/

File: subsys/power/device_pm.c (hunk 6 above)
Comment (ceolin) [reply subsys/power/device_pm.c:115]: yep, I tested with the blinky sample (last commit) and it seems to be working

File: drivers/pinmux/stm32/pinmux_stm32.c (hunk 7 above)
Comment (ceolin) [reply drivers/pinmux/stm32/pinmux_stm32.c:153]: yep, just changed to you be able to check / confirm

File: subsys/power/device.c (hunk 3 above)
Comment (ceolin) [reply subsys/power/device.c:108]: Nope, the first device that fails will cause this function to return with an error code. The system pm then will undo the changes calling `pm_resume_devices()` that will only iterate over the devices that have changed state.

File: drivers/gpio/gpio_stm32.c [hunk 8]
```diff
@@ -474,6 +474,12 @@ static int gpio_stm32_config(const struct device *dev,
 		goto exit;
 	}
 
+
+#ifdef CONFIG_PM_DEVICE_IDLE
```
Comment (gmarull) [inline drivers/gpio/gpio_stm32.c:478]: I wonder if it would be better to just add *dummy* PM functions when device PM is not enabled, so that we do not pollute drivers with tons of `#ifdef CONFIG_PM_XXX`. See, e.g. https://elixir.bootlin.com/linux/latest/source/include/linux/pm_runtime.h#L231 Opinions @ceolin @nashif ?

Comment (ceolin) [reply drivers/gpio/gpio_stm32.c:478]: I'm getting rid of CONFIG_PM_DEVICE_IDLE option, that should simplify all this `#ifdef` code. But I tend to agree with you about all these conditional code, we should avoid as much as possible. Dummy functions have to be carefully thought though, they will return an error code and these error codes have to be properly handled by the caller.

File: drivers/gpio/gpio_stm32.c [hunk 9]
```diff
@@ -561,6 +575,63 @@ static const struct gpio_driver_api gpio_stm32_driver = {
...
+
+	if (new_state == DEVICE_PM_ACTIVE_STATE) {
+		ret = gpio_stm32_clock_request(dev, true);
+	} else if (new_state == DEVICE_PM_SUSPEND_STATE) {
+		ret = gpio_stm32_clock_request(dev, false);
+	}
+
+	data->power_state = new_state;
+
+	return ret;
```
Comment (gmarull) [inline drivers/gpio/gpio_stm32.c:600]: should be changed to:
```c
if (ret < 0) {
    return ret;
}

data->power_state = new_state;

return 0;
```

File: drivers/gpio/gpio_stm32.c [hunk 10]
```diff
@@ -561,6 +575,63 @@ static const struct gpio_driver_api gpio_stm32_driver = {
...
+		if (new_state != data->power_state) {
+			ret = gpio_stm32_set_power_state(dev, new_state);
+		}
+		break;
+	case DEVICE_PM_GET_POWER_STATE:
+		*((uint32_t *)context) = gpio_stm32_get_power_state(dev);
+		break;
+	default:
+		ret = -EINVAL;
+
```
Comment (gmarull) [inline drivers/gpio/gpio_stm32.c:623]: nit: extra blank line not needed

File: subsys/power/device.c [hunk 11]
```diff
@@ -60,26 +60,65 @@ static device_idx_t num_pm;
...
 static device_idx_t num_susp;
 
+static bool should_suspend(const struct device *dev, uint32_t state)
+{
+	int rc;
+	uint32_t current_state;
+
+	rc = device_get_power_state(dev, &current_state);
+	if ((rc != -ENOTSUP) && (rc != 0)) {
+		LOG_DBG("Was not possible to get device %s state: %d",
```
Comment (gmarull) [inline subsys/power/device.c:70]: nit: `LOG_ERR`

File: subsys/power/device_pm.c [hunk 12]
```diff
@@ -157,6 +161,26 @@ int device_pm_put_sync(const struct device *dev)
...
+		int ret;
+
+		/*
+		 * If you are running in pre-kernel we don't need to
+		 * worry about multi threading.
+		 */
+		ret = device_set_power_state(dev,
+					     DEVICE_PM_ACTIVE_STATE,
+					     device_pm_callback, NULL);
+		if (ret == 0)  {
```
Comment (gmarull) [inline subsys/power/device_pm.c:170]: we should likely assert here, otherwise error condition will be silenced.

File: doc/reference/power_management/index.rst [hunk 13]
```diff
@@ -492,10 +492,6 @@ the following configuration flags.
    This flag is enabled if the SOC interface and the devices support device power
    management.
 
-:code:`CONFIG_PM_DEVICE_IDLE`
```
Comment (gmarull) [inline doc/reference/power_management/index.rst:495]: Comparing to Linux, this flag enables what they call "runtime PM". It may still be useful to have a distinction between device PM (used by system PM) and runtime device PM. In some cases we may want to have one but not the other, e.g. I want that all devices are suspended when I suspend the sytem, but I don't care about PM when on running (all `device_pm_get/put` should be a nop). We can discuss this again tomorrow.

File: subsys/power/device_pm.c [hunk 14]
```diff
@@ -192,9 +191,9 @@ void device_pm_enable(const struct device *dev)
 		dev->pm->dev = dev;
 		atomic_set(&dev->pm->fsm_state,
 			   DEVICE_PM_STATE_SUSPENDED);
-		k_work_init(&dev->pm->work, pm_work_handler);
+		k_work_init_delayable(&dev->pm->work, pm_work_handler);
 	} else {
-		k_work_submit(&dev->pm->work);
+		k_work_schedule(&dev->pm->work, K_USEC(0));
```
Comment (gmarull) [inline subsys/power/device_pm.c:193]: `K_NO_WAIT`

File: subsys/power/device_pm.c [hunk 15]
```diff
@@ -116,25 +121,19 @@ static int device_pm_request(const struct device *dev,
 		return 0;
 	}
 
-	k_work_submit(&dev->pm->work);
+	(void)k_work_schedule(&dev->pm->work, K_USEC(0));
```
Comment (gmarull) [inline subsys/power/device_pm.c:121]: `K_NO_WAIT`

File: subsys/power/device_pm.c [hunk 16]
```diff
@@ -207,6 +206,6 @@ void device_pm_disable(const struct device *dev)
 	k_sem_take(&dev->pm->lock, K_FOREVER);
 	dev->pm->enable = false;
 	/* Bring up the device before disabling the Idle PM */
-	k_work_submit(&dev->pm->work);
+	k_work_schedule(&dev->pm->work, K_USEC(0));
```
Comment (gmarull) [inline subsys/power/device_pm.c:206]: `K_NO_WAIT`

File: subsys/power/device_pm.c [hunk 17]
```diff
@@ -8,6 +8,7 @@
 #include <kernel.h>
 #include <device.h>
 #include <sys/__assert.h>
+#include <sys/check.h>
```
Comment (gmarull) [inline subsys/power/device_pm.c:11]: why is this needed?

File: subsys/power/device_pm.c [hunk 18]
```diff
@@ -18,27 +18,19 @@ LOG_MODULE_DECLARE(power);
...
-	DEVICE_PM_STATE_SUSPENDED,
-	DEVICE_PM_STATE_SUSPENDING,
-	DEVICE_PM_STATE_RESUMING,
-};
+/*
+ * Lets just re-use device pm states for FSM but add these two
+ * transitioning states for internal usage.
+ */
+#define DEVICE_PM_RESUMING_STATE   (DEVICE_PM_OFF_STATE + 1U)
+#define DEVICE_PM_SUSPENDING_STATE (DEVICE_PM_RESUMING_STATE + 1U)
```
Comment (gmarull) [inline subsys/power/device_pm.c:26]: I'd not define internal states based on the value of external ones, a change on the external side may create unwanted value clashes.

**Review by gmarull — COMMENTED** [review]
Added some more comments/questions

File: subsys/power/device_pm.c (hunk 14 above)
Comment (ceolin) [reply subsys/power/device_pm.c:193]: will change, that is the part that I'm not particularly happy, I have to use `k_work_delayable`  otherwise the thread is scheduled out and the notification is missed causing a dead lock. For now this seems to work, but I want to remove the need of the work queue.

File: subsys/power/device_pm.c (hunk 17 above)
Comment (ceolin) [reply subsys/power/device_pm.c:11]: leftover from a check I had added but end up removing.

File: subsys/power/device_pm.c (hunk 18 above)
Comment (ceolin) [reply subsys/power/device_pm.c:26]: I will unify this with that states in near future. But there was already a dependency between those states, but was very confuse having `DEVICE_PM_ACTIVE_STATE` and `DEVICE_PM_STATE_ACTIVE` :/

File: subsys/power/device.c [hunk 19]
```diff
@@ -60,26 +60,65 @@ static device_idx_t num_pm;
...
-		++num_susp;
+		/*
+		 * Just mark as suspended devices that were suspended now
+		 * otherwise we will resume devices that were already suspended
+		 * and not being used.
+		 * This still not optimal, since we are not distinguishing
+		 * between other states like DEVICE_PM_LOW_POWER_STATE.
+		 */
+		if (suspend) {
+			++num_susp;
```
Comment (nashif) [inline subsys/power/device.c:120]: why not fold into previous if?

Comment (ceolin) [reply subsys/power/device.c:120]: There is no reason looking it. Will put it inside the previous branch.

File: subsys/power/device.c (hunk 11 above)
Comment (ceolin) [reply subsys/power/device.c:70]: I can raise it to WRN but I don't think it should be ERR. ERR should be used for things that is not possible to recover

File: drivers/serial/uart_stm32.c [hunk 20]
```diff
@@ -1583,7 +1583,7 @@ DEVICE_DT_INST_DEFINE(index,						\
 		    &uart_stm32_init,					\
 		    &uart_stm32_pm_control,				\
 		    &uart_stm32_data_##index, &uart_stm32_cfg_##index,	\
-		    PRE_KERNEL_1, CONFIG_KERNEL_INIT_PRIORITY_DEVICE,	\
+		    POST_KERNEL, CONFIG_KERNEL_INIT_PRIORITY_DEVICE,	\
```
[the remark this answers is not in the document]
Comment (erwango) [reply drivers/serial/uart_stm32.c:1586]: @gmarull Yes. But this change is not meant to be merged. The only purpose was to be able to have a working configuration the time the underlying limitation is fixed. I think we'll soon be able to get rid of it.
